osu.Game/Skinning/Skin.cs

```diff
@@ -293,20 +304,71 @@ private void applyMigration(SkinLayoutInfo layout, GlobalSkinnableContainers tar
                         c.Type.Name == nameof(DefaultComboCounter) ||
                         c.Type.Name == nameof(ArgonComboCounter)).ToArray();
 
-                    layout.Update(null, globalHUDComponents.Except(comboCounters).ToArray());
+                    hudLayout.Update(null, globalHUDComponents.Except(comboCounters).ToArray());
 
                     resources.RealmAccess.Run(r =>
                     {
                         foreach (var ruleset in r.All<RulesetInfo>())
                         {
-                            layout.Update(ruleset, layout.TryGetDrawableInfo(ruleset, out var rulesetHUDComponents)
+                            hudLayout.Update(ruleset, hudLayout.TryGetDrawableInfo(ruleset, out var rulesetHUDComponents)
                                 ? rulesetHUDComponents.Concat(comboCounters).ToArray()
                                 : comboCounters);
                         }
                     });
 
                     break;
                 }
+
+                case 2:
+                {
+                    // Legacy health bar is moved from the global HUD components target into per-ruleset.
+                    // On osu!mania, the health bar is moved from the HUD to Playfield target for stage-based positioning.
+                    if (resources == null)
+                        break;
+
+                    var hudLayout = layoutInfos.GetValueOrDefault(GlobalSkinnableContainers.MainHUDComponents);
+                    var playfieldLayout = layoutInfos.GetValueOrDefault(GlobalSkinnableContainers.Playfield);
+
+                    if (hudLayout == null || !hudLayout.TryGetDrawableInfo(null, out var globalHUDComponents))
+                        globalHUDComponents = Array.Empty<SerialisedDrawableInfo>();
+
+                    var legacyHealthBars = globalHUDComponents.Where(h => h.Type.Name == nameof(LegacyHealthDisplay)).ToArray();
```

Comment:
Why is this checking the _name_ of the type _if you already have the type itself in hand_?

```suggestion
                    var legacyHealthBars = globalHUDComponents.Where(h => h.Type == typeof(LegacyHealthDisplay)).ToArray();
```

Reply:
Sure, I don't know how `Type` equality works, and I hope that if it does other stupid things like assembly version or whatever then it won't break this logic.

Reply:
I will assume it only checks the name. I tried stepping through the equality implementation and ended up with both sides being equal by reference, I don't know.

Reply:
If you need clarification here, given that `CreateInstance` works with however we're serialising types, this case also has to work. We do seem to be serialising versioning out, but I guess it works fine.